CIP-0037/CIP-0037.md

```diff
@@ -0,0 +1,352 @@
+* Non-degeneracy: e != 1
+* Computability: There exists an efficient algorithm to compute e
+
+where G1, G2 and GT are three distinct groups of order a prime q. Given that all three groups have the same 
+order, we can refer to the scalars of each group using the same type/structure. 
+
+## Specification
+We now provide the technical specification. 
+
+### Names and types/kinds for the new functions or types
+The added types will be the following, all of which can be represented as a byte array. Even if these types are 
+equivalent to byte arrays of a given size, it makes sense to include these types, to enforce deserialization, and 
+therefore some checks on the data used by the smart contract. In particular, all three types can only be achieved 
+with byte arrays that represent a point which is part of the prime order subgroup. 
+
+* `BLS12_381G1Element`
+* `BLS12_381G2Element`
+* `BLS12_381GTElement`
+
+We need to support the binary operation of G1 and G2 (which are additive groups), as well as the binary operation 
+over GT (which is represented as a multiplicative group). We also want to enable hashing to G1 and G2.
+In particular, we expose the hash to curve (which we denote with `H2G1` and `H2G2` for G1 and G2 
+respectively) algorithm as described in 
+[hash to curve draft](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve#section-8), 
+using `BLS12381G1_XMD:SHA-256_SSWU_RO_` and `BLS12381G2_XMD:SHA-256_SSWU_RO_` for G1 and G2 respectively.
+We do not include the type of the Scalar, nor operations related to it. These type of operations can already be 
+performed with the Integral type. In particular, we need the following functions: 
+
+* Group operations: 
+    * `BLS12_381_G1_add :: BLS12_381G1Element -> BLS12_381G1Element -> BLS12_381G1Element`
+    * `BLS12_381_G1_mult :: Integer -> BLS12_381G1Element -> BLS12_381G1Element`
+    * `BLS12_381_G1_neg :: BLS12_381G1Element -> BLS12_381G1Element`
+    * `BLS12_381_H2G1 :: ByteArray -> BLS12_381G1Element`
+    * `BLS12_381_G2_add :: BLS12_381G2Element -> BLS12_381G2Element -> BLS12_381G2Element`
+    * `BLS12_381_G2_mult :: Integer -> BLS12_381G2Element -> BLS12_381G2Element`
+    * `BLS12_381_G2_neg :: BLS12_381G2Element -> BLS12_381G2Element`
+    * `BLS12_381_H2G2 :: ByteArray -> BLS12_381G2Element`
+    * `BLS12_381_GT_mul :: Blst12381GTElement -> Blst12381GTElement -> Blst12381GTElement`
+* Pairing operations:
+    * `BLS12_381_pairing :: BLS12_381G1Element -> BLS12_381G2Element -> BLS12_381GTElement`
+    * `BLS12_381_final_verify :: BLS12_381GTElement -> BLS12_381GTElement -> bool` This performs the final 
+  exponantiation (see section `An important note on GT elements` below).
+
+On top of the elliptic curve operations, we also need to include deserialization functions, and equality definitions
+among the G1 and G2 types. 
+
+* Deserialisation: 
+  * `deserializeG1 :: ByteString -> (BLSTError -> a) -> (Bls12_381G1Element -> a) -> a`
```

Comment:
Why not just `deserialiseG1 :: ByteString -> BuiltinList Bls12_381G1Element`, where the output is empty if it errs and otherwise a single element? We don't really care about error handling too much on-chain.
You could also just make it partial, `deserialiseG1 :: ByteString -> Bls12_381G1Element` looks reasonable to me.

Reply:
I can't think of a single case where handling the error is necessary.

Reply:
Why can't we just fail the whole thing if deserialization fails? Surely this indicates malformed input of some kind?

Reply:
I am very much for just failing the whole thing.